experimental/hip/hip_allocator.c

```diff
@@ -0,0 +1,586 @@
+            IREE_STATUS_INVALID_ARGUMENT,
+            "unable to register host allocations as device-local memory");
+      }
+      buffer_type = IREE_HAL_HIP_BUFFER_TYPE_HOST_REGISTERED;
+      host_ptr = external_buffer->handle.host_allocation.ptr;
+      uint32_t register_flags = hipHostRegisterMapped;
+      status = IREE_HIP_RESULT_TO_STATUS(
+          allocator->symbols,
+          hipHostRegister(host_ptr, external_buffer->size, register_flags),
+          "hipHostRegister");
+      if (iree_status_is_ok(status)) {
+        status = IREE_HIP_RESULT_TO_STATUS(
+            allocator->symbols,
+            hipHostGetDevicePointer(&device_ptr, host_ptr, 0),
+            "hipHostGetDevicePointer");
+      }
+      break;
+    }
+    case IREE_HAL_EXTERNAL_BUFFER_TYPE_DEVICE_ALLOCATION: {
+      buffer_type = IREE_HAL_HIP_BUFFER_TYPE_EXTERNAL;
+      device_ptr =
+          (hipDeviceptr_t)external_buffer->handle.device_allocation.ptr;
+      break;
+    }
+    case IREE_HAL_EXTERNAL_BUFFER_TYPE_OPAQUE_FD:
+    case IREE_HAL_EXTERNAL_BUFFER_TYPE_OPAQUE_WIN32:
+      return iree_make_status(IREE_STATUS_UNIMPLEMENTED,
+                              "handle-based imports not yet implemented");
+    default:
+      return iree_make_status(IREE_STATUS_UNIMPLEMENTED,
+                              "external buffer type not supported");
+  }
+
+  iree_hal_buffer_t* buffer = NULL;
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_hip_buffer_wrap(
+        base_allocator, compat_params.type, compat_params.access,
+        compat_params.usage, external_buffer->size,
+        /*byte_offset=*/0,
+        /*byte_length=*/external_buffer->size, buffer_type, device_ptr,
+        host_ptr, release_callback,
+        iree_hal_allocator_host_allocator(base_allocator), &buffer);
+  }
+
+  if (iree_status_is_ok(status)) {
+    *out_buffer = buffer;
+  } else {
+    if (!buffer && (device_ptr || host_ptr)) {
+      iree_hal_hip_buffer_free(allocator->symbols, buffer_type, device_ptr,
+                               host_ptr);
+    } else {
+      iree_hal_buffer_release(buffer);
+    }
+  }
+  return status;
+}
+
+static iree_status_t iree_hal_hip_allocator_export_buffer(
+    iree_hal_allocator_t* IREE_RESTRICT base_allocator,
+    iree_hal_buffer_t* IREE_RESTRICT buffer,
+    iree_hal_external_buffer_type_t requested_type,
+    iree_hal_external_buffer_flags_t requested_flags,
+    iree_hal_external_buffer_t* IREE_RESTRICT out_external_buffer) {
+  iree_hal_hip_buffer_type_t buffer_type = iree_hal_hip_buffer_type(buffer);
+
+  switch (requested_type) {
+    case IREE_HAL_EXTERNAL_BUFFER_TYPE_DEVICE_ALLOCATION:
+      switch (buffer_type) {
+        case IREE_HAL_HIP_BUFFER_TYPE_EXTERNAL:
+          out_external_buffer->flags = requested_flags;
+          out_external_buffer->type = requested_type;
+          out_external_buffer->handle.device_allocation.ptr =
+              *((uint64_t*)iree_hal_hip_buffer_device_pointer(buffer));
```

Comment:
We should do `(uint64_t)(uintptr_t)` to cast the return value to a `uint64_t` here.